### PR TITLE
geopm-service: New package and deprecate geopm

### PR DESCRIPTION
**Pull request by bgeltz**
- Add geopm-service and its dependencies: py-sphinx-emoji and py-dasbus.
- Deprecate old geopm package as it is no longer supported.
- GEOPM is now distributed as 2 packages: geopm-service and geopm-runtime.
- geopm-runtime will be dependent on geopm-service (in particular libgeopmd.so.2).  A PR for geopm-runtime is forthcoming.
- Relates to #26003.
- Relates to #38795.
- Relates to #38798.

### CODE REVIEW

**Comment by bgeltz** [comment]
I'm posting this PR to get some early feedback, but I'm still performing final checks on the spack installed geopm-service package.  Thanks!

File: var/spack/repos/builtin/packages/geopm/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm/package.py:1]: Normally, we first deprecate all versions and wait a full release cycle before removing versions/packages: https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions

File: var/spack/repos/builtin/packages/geopm-service/package.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    git = "https://github.com/geopm/geopm.git"
+
+    maintainers("bgeltz", "cmcantalupo")
+    license("BSD-3-Clause")
+    tags = ["e4s"]
+
+    version("develop", branch="dev", get_full_repo=True)
+    version(
+        "3.0.1",
+        url="https://github.com/geopm/geopm/tarball/v3.0.1",
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm-service/package.py]: Could just move this URL up to the package level

File: var/spack/repos/builtin/packages/geopm-service/package.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    depends_on("doxygen", type="build", when="+docs")
+    depends_on("py-docstring-parser@0.13.0:", type="build", when="+docs")
+    depends_on("py-sphinx@4.5:", type="build", when="+docs")
+    depends_on("py-sphinx-rtd-theme@1:", type="build", when="+docs")
+    depends_on("py-sphinx-emoji@0.2.0:", type="build", when="+docs")
+    depends_on("py-sphinx-tabs@3.3.1:", type="build", when="+docs")
+    depends_on("py-pygments@2.13.0:", type="build", when="+docs")
+
+    # Other Python dependencies
+    depends_on("python@3.6:3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm-service/package.py]: Python isn't needed at run-time?

I'm guessing these deps come from `service/setup.py`, can you add a comment to make them easier to find?

File: var/spack/repos/builtin/packages/geopm-service/package.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    def setup_run_environment(self, env):
+        if os.path.isdir(self.prefix.lib64):
+            lib_dir = self.prefix.lib64
+        else:
+            lib_dir = self.prefix.lib
+        env.prepend_path("LD_LIBRARY_PATH", lib_dir)
+
+        python_version = self.spec["python"].version.up_to(2)
+        python_inst_dir = join_path(lib_dir, "python{0}".format(python_version), "site-packages")
+        env.prepend_path("PYTHONPATH", python_inst_dir)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm-service/package.py]: If you add:
```python
extends("python")
```
all of this will be done automatically for you, and it will work even if the Python libs are installed to a different location.

File: var/spack/repos/builtin/packages/py-dasbus/package.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class PyDasbus(PythonPackage):
+    """Dasbus is a DBus library written in Python 3, based on GLib and inspired by pydbus."""
+
+    homepage = "https://dasbus.readthedocs.io/en/latest/"
+    pypi = "dasbus/dasbus-1.7.tar.gz"
+    license("LGPL-2.1-or-later")
+
+    version("1.7", sha256="a8850d841adfe8ee5f7bb9f82cf449ab9b4950dc0633897071718e0d0036b6f6")
+
+    extends("python")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dasbus/package.py]: This is automatically added by `PythonPackage`, you don't need to add it.

File: var/spack/repos/builtin/packages/py-dasbus/package.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    depends_on("python@3.6:3", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+    depends_on("py-pygobject", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dasbus/package.py:20]: I see this listed in the README, but I don't actually see where the package uses it...

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 6]
```diff
@@ -10,8 +10,10 @@ class PyPygobject(PythonPackage):
     """bindings for the GLib, and GObject,
     to be used in Python."""
 
-    homepage = "https://pypi.python.org/pypi/pygobject"
+    homepage = "https://pygobject.readthedocs.io/en/latest/"
+    pypi = "PyGObject/PyGObject-3.46.0.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]: This isn't actually being used, correct?

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 7]
```diff
@@ -10,8 +10,10 @@ class PyPygobject(PythonPackage):
...
 
+    version("3.46.0", sha256="426008b2dad548c9af1c7b03b59df0440fde5c33f38fb5406b103a43d653cafc")
     version("3.38.0", sha256="0372d1bb9122fc19f500a249b1f38c2bb67485000f5887497b4b205b3e7084d5")
     version("3.28.3", sha256="3dd3e21015d06e00482ea665fc1733b77e754a6ab656a5db5d7f7bfaf31ad0b0")
     version("2.28.6", sha256="fb8a1d4f665130a125011659bd347c7339c944232163dbb9a34fd0686577adb8")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]: Should probably update the deps, setup.py now has a list of required versions

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+class PySphinxEmoji(PythonPackage):
+    """An extension to use emoji codes in your Sphinx documentation!"""
+
+    homepage = "https://sphinxemojicodes.readthedocs.io"
+    pypi = "sphinxemoji/sphinxemoji-0.2.0.tar.gz"
+    license("BSD-3-Clause")
+
+    version("0.2.0", sha256="27861d1dd7c6570f5e63020dac9a687263f7481f6d5d6409eb31ecebcc804e4c")
+
+    depends_on("python@3.6:3", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-emoji/package.py]: Where do you see this?

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    depends_on("py-setuptools", type="build")
+    depends_on("py-sphinx@2.1:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-emoji/package.py]:
```suggestion
    depends_on("py-sphinx@1.8:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py [hunk 10]
```diff
@@ -15,6 +15,10 @@ class PySphinxTabs(PythonPackage):
 
     maintainers("schmitts")
 
+    version("3.4.4", sha256="f1b72c4f23d1ba9cdcaf880fd883524bc70689f561b9785719b8b3c3c5ed0aca")
+    version("3.4.1", sha256="d2a09f9e8316e400d57503f6df1c78005fdde220e5af589cc79d493159e1b832")
+    version("3.4.0", sha256="75e97ce10b74700deaf87b662539a293c8afc9dfa9d21f126b860118064cb0c5")
+    version("3.3.1", sha256="d10dd7fb2700329b8e5948ab9f8e3ef54fff30f79d2e42cfd1b0089ae26e8c5e")
     version("3.2.0", sha256="33137914ed9b276e6a686d7a337310ee77b1dae316fdcbce60476913a152e0a4")
 
     depends_on("python@3.6:3", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-tabs/package.py:25]: Need to update deps:
```python
    install_requires=["sphinx", "pygments", "docutils~=0.18.0"],                         
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-dasbus/package.py (hunk 5 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-dasbus/package.py:20]: AFAICT, the dasbus client and server implementations use pygobject through the "gi" API.

File: var/spack/repos/builtin/packages/py-dasbus/package.py (hunk 4 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-dasbus/package.py]: Got it.  I'll remove.

File: var/spack/repos/builtin/packages/geopm-service/package.py (hunk 2 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/geopm-service/package.py]: Python is needed at runtime.  I'll update and add a comment.

File: var/spack/repos/builtin/packages/geopm/package.py
Comment (bgeltz) [reply var/spack/repos/builtin/packages/geopm/package.py:1]: I was hoping to avoid any confusion for new users given that the older "geopm" recipe was a monolithic package while the new recipes will be split between the "geopm-service" and "geopm-runtime".  The old recipe should not be used; it's currently broken on most systems.

I'll add it back and mark it all as deprecated.

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py (hunk 8 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-sphinx-emoji/package.py]: Looking at their GitHub:
https://github.com/sphinx-contrib/emojicodes/blob/master/setup.py
Calls out sphinx >= 1.8.
Sphinx @ 1.8 supported back to Python 3.4:
https://github.com/sphinx-doc/sphinx/blob/v1.8.0/setup.py

I'll update and test.

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py (hunk 10 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-sphinx-tabs/package.py:25]: I'll update the depends below for this.

File: var/spack/repos/builtin/packages/geopm-service/package.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    variant(
+        "rawmsr",
+        default=True,
+        description="Enable direct use of standard msr device driver",
+        when="@develop",
+    )
+
+    conflicts("+nvml", when="+levelzero", msg="LevelZero and NVML support are mutually exclusive")
+    conflicts("+levelzero", when="+nvml", msg="LevelZero and NVML support are mutually exclusive")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm-service/package.py]: No need to duplicate this, they are symmetric.

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 12]
```diff
@@ -10,22 +10,38 @@ class PyPygobject(PythonPackage):
...
     depends_on("pkgconfig", type="build")
-    depends_on("libffi")
-    depends_on("glib")
-    depends_on("py-pycairo", type=("build", "run"))
-    depends_on("gobject-introspection")
     depends_on("gtkplus")
 
+    depends_on("libffi@3.0.0:", when="@3.46.0")
+    depends_on("glib@2.64.0:", when="@3.46.0")
+    depends_on("py-pycairo@1.16.0:", type=("build", "run"), when="@3.46.0")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]:
```suggestion
    depends_on("py-pycairo@1.16:1", type=("build", "run"), when="@3.46.0")
```

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 13]
```diff
@@ -10,22 +10,38 @@ class PyPygobject(PythonPackage):
...
 
     extends("python")
 
     depends_on("py-setuptools", type="build")
     depends_on("pkgconfig", type="build")
-    depends_on("libffi")
-    depends_on("glib")
-    depends_on("py-pycairo", type=("build", "run"))
-    depends_on("gobject-introspection")
     depends_on("gtkplus")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygobject/package.py:39]: Needs `python@3.8:3` dep

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pygobject/package.py:39]: Can you add a comment like:
```python
# meson.build
```
so it's obvious where to look to find dependency versions? Would also help to sort them in the same order.

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+class PySphinxEmoji(PythonPackage):
+    """An extension to use emoji codes in your Sphinx documentation!"""
+
+    homepage = "https://sphinxemojicodes.readthedocs.io"
+    pypi = "sphinxemoji/sphinxemoji-0.2.0.tar.gz"
+    license("BSD-3-Clause")
+
+    version("0.2.0", sha256="27861d1dd7c6570f5e63020dac9a687263f7481f6d5d6409eb31ecebcc804e4c")
+
+    depends_on("python@3.4:3", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-emoji/package.py]: Where do you see this?

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py [hunk 15]
```diff
@@ -15,10 +15,16 @@ class PySphinxTabs(PythonPackage):
...
 
+    version("3.4.4", sha256="f1b72c4f23d1ba9cdcaf880fd883524bc70689f561b9785719b8b3c3c5ed0aca")
+    version("3.4.1", sha256="d2a09f9e8316e400d57503f6df1c78005fdde220e5af589cc79d493159e1b832")
+    version("3.3.1", sha256="d10dd7fb2700329b8e5948ab9f8e3ef54fff30f79d2e42cfd1b0089ae26e8c5e")
     version("3.2.0", sha256="33137914ed9b276e6a686d7a337310ee77b1dae316fdcbce60476913a152e0a4")
 
     depends_on("python@3.6:3", type=("build", "run"))
     depends_on("py-setuptools", type="build")
-    depends_on("py-sphinx@2:4", type=("build", "run"))
+    depends_on("py-sphinx@2:7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-tabs/package.py]: The latest version claims to work for any Sphinx version, but not for older versions, so neither of these are right.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py (hunk 14 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-sphinx-emoji/package.py]: I had inferred this because of the depends_on for sphinx 1.8.  I'll drop it from here.

File: var/spack/repos/builtin/packages/geopm-service/package.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    depends_on("doxygen", type="build", when="+docs")
+    depends_on("py-docstring-parser@0.13.0:", type="build", when="+docs")
+    depends_on("py-sphinx@4.5:", type="build", when="+docs")
+    depends_on("py-sphinx-rtd-theme@1:", type="build", when="+docs")
+    depends_on("py-sphinx-emoji@0.2.0:", type="build", when="+docs")
+    depends_on("py-sphinx-tabs@3.3.1:", type="build", when="+docs")
+    depends_on("py-pygments@2.13.0:", type="build", when="+docs")
+
+    # Other Python dependencies - from service/setup.py
+    depends_on("python@3.6:3", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/geopm-service/package.py]:
```suggestion
    depends_on("python@3.6:", type=("build", "run"))
```
Could just remove this line, Spack only supports Python 3.8+ anyway

File: var/spack/repos/builtin/packages/gtkplus/package.py [hunk 17]
```diff
@@ -36,7 +36,7 @@ class Gtkplus(MesonPackage):
         deprecated=True,
     )
 
-    variant("cups", default=False, description="enable cups support")
+    variant("cups", default=True, description="enable cups support")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gtkplus/package.py]: Why this change? If another package requires `gtkplus+cups`, it should say that.

File: var/spack/repos/builtin/packages/py-sphinx-emoji/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-emoji/package.py:1]: Shouldn't this be named `py-sphinxemoji`?

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py [hunk 18]
```diff
@@ -15,10 +15,17 @@ class PySphinxTabs(PythonPackage):
...
     depends_on("py-setuptools", type="build")
-    depends_on("py-sphinx@2:4", type=("build", "run"))
+    depends_on("py-sphinx", type=("build", "run"), when="@3.4.1:")
+    depends_on("py-sphinx@2:5", type=("build", "run"), when="@:3.4.1")
     depends_on("py-pygments", type=("build", "run"))
-    depends_on("py-docutils@0.16", type=("build", "run"))
+
+    depends_on("py-docutils@0.16:", when="@3.2.0", type=("build", "run"))
+    depends_on("py-docutils@0.17:", when="@3.3.1", type=("build", "run"))
+    depends_on("py-docutils@0.18:", when="@3.4.1:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-tabs/package.py]:
```suggestion
    depends_on("py-docutils@0.16", when="@3.2.0", type=("build", "run"))
    depends_on("py-docutils@0.17", when="@3.3.1", type=("build", "run"))
    depends_on("py-docutils@0.18", when="@3.4.1:", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py [hunk 19]
```diff
@@ -15,10 +15,17 @@ class PySphinxTabs(PythonPackage):
...
+    version("3.4.4", sha256="f1b72c4f23d1ba9cdcaf880fd883524bc70689f561b9785719b8b3c3c5ed0aca")
+    version("3.4.1", sha256="d2a09f9e8316e400d57503f6df1c78005fdde220e5af589cc79d493159e1b832")
+    version("3.3.1", sha256="d10dd7fb2700329b8e5948ab9f8e3ef54fff30f79d2e42cfd1b0089ae26e8c5e")
     version("3.2.0", sha256="33137914ed9b276e6a686d7a337310ee77b1dae316fdcbce60476913a152e0a4")
 
     depends_on("python@3.6:3", type=("build", "run"))
     depends_on("py-setuptools", type="build")
-    depends_on("py-sphinx@2:4", type=("build", "run"))
+    depends_on("py-sphinx", type=("build", "run"), when="@3.4.1:")
+    depends_on("py-sphinx@2:5", type=("build", "run"), when="@:3.4.1")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-sphinx-tabs/package.py]:
```suggestion
    depends_on("py-sphinx@2:4", type=("build", "run"), when="@:3.3.1")
```
Version ranges in Spack are inclusive

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gtkplus/package.py (hunk 17 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/gtkplus/package.py]: In my testing, ```spack install gtkplus``` will fail unless your build host has cups.h available.  It seems the build process for gtkplus has a hidden dependency on cups even if the ~cups variant is the default.  
My build logs are here: https://gist.github.com/bgeltz/d33f9198a74ff0f0047766ab5ec7df02

Alternatively, I can try to build a newer version of gtkplus than 3.24.29 to see if it is resolved, but I had not yet gone down that path.  I see on gtk.org their "latest old stable" is 3.24.39.

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py (hunk 18 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/py-sphinx-tabs/package.py]: The setup.py's all use ```~=``` to specify the py-docutils version that is required for the desired sphinx-tabs version.   AFAICT, that ```~=``` means ```>=``` to the specified version as long as the major version numbers match.  Since py-docutils was still on version 0, I thought I had captured this properly, or perhaps it should have been like ```depends_on("py-docutils@0.16:0"...```

Is there something special about docutils that would require pinned versions as you requested?

File: var/spack/repos/builtin/packages/gtkplus/package.py (hunk 17 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/gtkplus/package.py]: Let's separate that out into a separate PR. Changing the default doesn't fix the issue, it just hides it.

File: var/spack/repos/builtin/packages/py-sphinx-tabs/package.py (hunk 18 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-sphinx-tabs/package.py]: See https://peps.python.org/pep-0440/#compatible-release, `~= 0.18.0` means `>= 0.18.0, == 0.18.*`

File: var/spack/repos/builtin/packages/gtkplus/package.py (hunk 17 above)
Comment (bgeltz) [reply var/spack/repos/builtin/packages/gtkplus/package.py]: Done.  It's here: https://github.com/spack/spack/pull/42301

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
Need to resolve merge conflicts

**Review by adamjstewart — APPROVED** [review]
